cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/query/IndexCreator.java

```diff
@@ -355,4 +371,47 @@ private String createVirtualTableStatementForIndex(String indexName,
                                                                               settings);
     }
 
+    /**
+     * Iterate candidate indexNames generated from the indexNameRandom generator
+     * until we find one which doesn't already exist.
+     *
+     * We make sure the generated name is not an index already by the list
+     * of index names returned by {@link #listIndexesInDatabaseQueue()} method.
+     * This is because we avoid knowing about how indexes are stored in SQLite, however
+     * this means that it is not thread safe, it is possible for a new index with the same
+     * name to be created after a copy of the indexes has been taken from the database.
+     *
+     * We allow up to 200 random name generations, which should give us many millions
+     * of indexes before a name fails to be generated and makes sure this method doesn't
+     * loop forever.
+     *
+     * @return The generated index name or {@code null} if it failed.
+     *
+     */
+    private String generateIndexName() throws ExecutionException, InterruptedException {
+
+        String indexName = null;
+        Set<String> indexNames = listIndexesInDatabaseQueue().keySet();
+        Hex hex = new Hex();
+
+        int tries = 0;
+        byte[] randomBytes = new byte[20];
+        while (tries < 200 && indexName == null) {
+            indexNameRandom.nextBytes(randomBytes);
+            String candidate = new String(hex.encode(randomBytes), Charset.forName("UTF-8"));
+
+            if(!indexNames.contains(candidate)){
+                indexName = candidate;
+            }
+            tries++;
+        }
+
+        if (indexName != null) {
+            return indexName;
+        } else {
+            logger.warning("Failed to generate unique index name");
```

Comment:
I'd be tempted to log this on line 122, which adds a bit of self-documentation to why we're failing out there.